File: 2024/soutez/03-Holub/solution/solution__.py


gb2mb = lambda x: 1024*x
mb2gb = lambda x: x/1024



if __name__ == '__main__':
    distance, capacity, velocity, landline_transfer_speed, file_n, file_size = [float(input()) for _ in range(6)]
    capacity = gb2mb(capacity)
    

    # maximum files we can fit on the homing pigeon
    hp_max_files = capacity//file_size

    # maximum gb capacity of pigeon
    hp_usable_capacity = hp_max_files*file_size # (Gb)

    # flight time of pigeon
    hp_flight_time = velocity/distance # (sec)
    
    
    # we need to find which option will take the longest
    # (pigeon travels parallely with landline data, one doesn't wait for the other to finish)
    max_t = 0 # (sec)


    # if homing pigeon has better transfer speed
    if gb2mb(hp_usable_capacity)/hp_flight_time <= landline_transfer_speed:
        # send maximum files over pigeon
        file_n -= hp_max_files
        # time for pigeon to reach endpoint
        max_t = hp_flight_time
        

    # time to send the rest of files over landline
    landline_time = gb2mb(file_n * file_size) / landline_transfer_speed # (sec)

    max_t = max(max_t, landline_time)

    print(int(max_t))

File: 2024/soutez/03-Holub/solution/solution.cpp
#include <iostream>

int main(){
    int d, c, v, n, s;
    float p;

    std::cin >> d >> c >> v >> p >> n >> s;

    float p_gb = p/1024.0; // mb/s to gb/s

    int pigeon_max_n = c / s; // maximum files

    float flight_time = (float)d / (float)v;

    float pigeon_transfer_speed = (float)(pigeon_max_n*s) / flight_time;

    bool pigeon_faster = false;
    if(pigeon_transfer_speed > p_gb){ // pigeon faster
        n -= pigeon_max_n;
        pigeon_faster = true;
    }

    float total_land_time = n*s / p_gb; // total size / landline transfer speed

    if(pigeon_faster && flight_time > total_land_time){ // flight takes longer
        std::cout << (int)flight_time << '\n';
    }
    else{ // pigeon not deployed or landline takes longer
        std::cout << (int)total_land_time << '\n';
    }
}